### PR TITLE
ci: CI_SANITIZER flag to control sanitizer usage

### PR DESCRIPTION
**Pull request by def-**
Can be enabled using https://trigger-ci.dev.materialize.com/ for each pipeline

Mostly tested with address sanitizer, which also found the most interesting issues, so this is what I will expose in trigger-ci for now.

The most hacky part is that ASan in combined Rust/C/C++ code requires building with Clang.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: misc/python/materialize/mzbuild.py
Comment (benesch) [inline misc/python/materialize/mzbuild.py:969]: Recommend using `choices` here to enforce valid options (and then you can avoid enumerating them by hand in the help string too).

**Comment by benesch** [comment]
Looks pretty good! Took a quick skim and left a few comments but I probably won't have time to do a full re-review and approval. Nice work getting this all wired up.

File: ci/mkpipeline.py
Comment (nrainer-materialize) [inline ci/mkpipeline.py:123]: Should we do this with coverage builds as well? Not with factor 3 but probably some `1.x`

**Review by nrainer-materialize — APPROVED** [review]

File: ci/mkpipeline.py
Comment (def-) [reply ci/mkpipeline.py:123]: We already do that for coverage with factor 3.

**Comment by def-** [comment]
Planning to merge this later today, so if there are any complaints, please tell me.

File: misc/python/materialize/rustc_flags.py
Comment (nrainer-materialize) [inline misc/python/materialize/rustc_flags.py:46]: Shall we call it `undefined_behavior`? (optional)

File: ci/mkpipeline.py
Comment (nrainer-materialize) [inline ci/mkpipeline.py:305]: Is `nightly` here correct and if so, why does it differ from `nightlies` in Buildkite?

Comment (nrainer-materialize) [inline ci/mkpipeline.py:305]: How were these pipelines chosen? I assume these are the testing pipelines. Should we add the SLT pipeline as well?

Comment (nrainer-materialize) [reply ci/mkpipeline.py:305]: Should we extract this set of values to a named constant in `buildkite_constants.py`? (optional)
(We can also move that file to a more generic place.)

**Review by nrainer-materialize — APPROVED** [review]
Just skimmed over it, LGTM besides the minor comments.

File: misc/python/materialize/rustc_flags.py
Comment (def-) [reply misc/python/materialize/rustc_flags.py:46]: The compiler flag is also `-fsanitize=undefined`, so I'd prefer to stay with the current one.

File: ci/mkpipeline.py
Comment (def-) [reply ci/mkpipeline.py:305]: Yes, it's a mess because the Buildkite pipeline is called nightlies, but the directory is ci/nightly. We should unify those by having a symbolic link for a while, then switching over to the new name.

Comment (def-) [reply ci/mkpipeline.py:305]: These are the pipelines which don't have a global `wait` block after the compilation, since there are jobs that can be run without waiting for all compilations to be done, for example triggering another pipeline, or running an x86-64 job before the arm build is finished.

**Pull request merged**
